File: src/main/java/com/qtech/bubbles/gui/PauseButton.java
package com.qtech.bubbles.gui;

import com.qtech.bubbles.QBubbles;
import com.qtech.bubbles.common.interfaces.Listener;
import com.qtech.bubbles.core.controllers.MouseController;
import com.qtech.bubbles.event.MouseEvent;
import com.qtech.bubbles.event.RenderEventPriority;
import com.qtech.bubbles.event.SubscribeEvent;
import com.qtech.bubbles.event.bus.EventBus;
import com.qtech.bubbles.event.type.MouseEventType;
import com.qtech.bubbles.graphics.Border;
import com.qtech.bubbles.util.Util;

import java.awt.*;
import java.util.Objects;

public class PauseButton extends Widget implements Listener {
    private final long hash;
    private Rectangle bounds;
    private RenderEventPriority renderEventPriority;
    private boolean hovered;
    private boolean pressed;
    private boolean eventsActive = false;
    private Runnable command;
    private int clickCount;
    private String text;
    private EventBus.Handler binding;

    public void setText(String text) {
        this.text = text;
    }

    public String getText() {
        return text;
    }

    public static class Builder {
        private Rectangle _bounds = new Rectangle(10, 10, 96, 48);
        private String _text = "";
        private RenderEventPriority _renderPriority = RenderEventPriority.AFTER_FILTER;
        private Runnable _command = () -> {
        };

        public Builder() {
        }

        public PauseButton build() {
            PauseButton button = new PauseButton(_bounds.x, _bounds.y, _bounds.width, _bounds.height, _renderPriority);

            button.setText(_text);
            button.setCommand(_command);
            return button;
        }

        public Builder bounds(Rectangle bounds) {
            this._bounds = bounds;
            return this;
        }

        public Builder bounds(int x, int y, int width, int height) {
            this._bounds = new Rectangle(x, y, width, height);
            return this;
        }

        public Builder text(String text) {
            this._text = text;
            return this;
        }

        public Builder renderPriority(RenderEventPriority renderPriority) {
            this._renderPriority = renderPriority;
            return this;
        }

        public Builder command(Runnable command) {
            this._command = command;
            return this;
        }
    }

    protected PauseButton(int x, int y, int width, int height) {
        this(x, y, width, height, RenderEventPriority.AFTER_FILTER);
    }

    protected PauseButton(int x, int y, int width, int height, RenderEventPriority renderEventPriority) {
        this.renderEventPriority = renderEventPriority;
        this.bounds = new Rectangle(x, y, width, height);

        hash = System.nanoTime();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        PauseButton that = (PauseButton) o;
        return hash == that.hash;
    }

    @Override
    public int hashCode() {
        return Objects.hash(hash);
    }

    @SubscribeEvent
    public void onMouse(MouseEvent evt) {
        if (evt.getType() == MouseEventType.PRESS) {
            if (bounds.contains(evt.getPoint())) {
                pressed = true;
            }
        } else if (evt.getType() == MouseEventType.RELEASE) {
            if (bounds.contains(evt.getPoint())) {
                command.run();
            }
            pressed = false;
//        } else if (evt.getType() == MouseEventType.CLICK) {
        }
    }

//    @SubscribeEvent
//    public void onMouseMotion(MouseMotionEvent evt) {
//        if (bounds.contains(evt.getParentEvent().getPoint())) {
//            QBubbles.setCursor(Game.instance().getPointerCursor());
//            hovered = true;
//        } else {
//            if (hovered) {
//                QBubbles.setCursor(Game.instance().getDefaultCursor());
//                hovered = false;
//            }
//        }
//    }

    public void paint(Graphics2D gg) {
//        hovered = MouseController.instance().getCurrentPoint() != null && bounds.contains(MouseController.instance().getCurrentPoint());

        Point mousePos = MouseController.instance().getCurrentPoint();
        boolean hoveredNew = mousePos != null && bounds.contains(mousePos);
        if (mousePos != null) {
            if (bounds.contains(mousePos)) {
                Util.setCursor(QBubbles.getInstance().getPointerCursor());
                hovered = true;
            } else {
                if (hovered) {
                    Util.setCursor(QBubbles.getInstance().getDefaultCursor());
                    hovered = false;
                }
            }
        }

        Color textColor;

        Stroke oldStroke = gg.getStroke();

        if (pressed && MouseController.instance().getCurrentPoint() != null && bounds.contains(MouseController.instance().getCurrentPoint())) {

            Paint old = gg.getPaint();
            GradientPaint p = new GradientPaint(0, bounds.y, new Color(0, 192, 255), 0f, bounds.y + bounds.height, new Color(0, 255, 192));
            gg.setPaint(p);
            Border border = new Border(1, 1, 1, 1);
            border.setPaint(new Color(255, 255, 255, 128));
            border.paintBorder(QBubbles.getInstance(), gg, bounds.x, bounds.y, bounds.width, bounds.height);
            gg.setPaint(old);

            Border border1 = new Border(0, 0, 2, 0);
            border1.setPaint(p);
            border1.paintBorder(QBubbles.getInstance(), gg, bounds.x, bounds.y, bounds.width, bounds.height);

            textColor = Color.white;
        } else if (hovered) {
            gg.setStroke(new BasicStroke(4.0f));

            Paint old = gg.getPaint();
            double shiftX = ((double)bounds.width * 2) * QBubbles.getTicks() / (QBubbles.TPS * 10);
            GradientPaint p = new GradientPaint(bounds.x + ((float) shiftX - bounds.width), 0, new Color(0, 192, 255), bounds.x + (float)shiftX, 0f, new Color(0, 255, 192), true);
            gg.setPaint(p);

            gg.setColor(new Color(255, 255, 255, 128));
            gg.fill(bounds);

            Border border1 = new Border(0, 0, 2, 0);
            border1.setPaint(p);
            border1.paintBorder(QBubbles.getInstance(), gg, bounds.x, bounds.y, bounds.width, bounds.height);
//            gg.draw(new Rectangle(bounds.x - 2, bounds.y - 2, bounds.width + 4, bounds.height + 4));

            gg.setPaint(old);

            textColor = new Color(255, 255, 255);
        } else {
            gg.setStroke(new BasicStroke(1.0f));

            gg.setColor(new Color(255, 255, 255, 128));
            gg.fill(bounds);
//            Border border = new Border(1, 1, 1, 1);
//            border.setPaint(new Color(255, 255, 255, 128));
//            border.paintBorder(QBubbles.getInstance(), gg, bounds.x, bounds.y, bounds.width, bounds.height);
//            gg.draw(new Rectangle(bounds.x + 1, bounds.y + 1, bounds.width - 2, bounds.height - 2));

            textColor = new Color(255, 255, 255, 128);
        }

        OptionsNumberInput.NumberInputButton.paint0a(gg, textColor, oldStroke, bounds, text);
    }

    @Override
    public void destroy() {

    }

    @SuppressWarnings("EmptyMethod")
    public void tick() {

    }

    @Override
    public void bindEvents() {
        QBubbles.getEventBus().register(this);
        eventsActive = true;

        Point mousePos = MouseController.instance().getCurrentPoint();
        boolean hoveredNew = mousePos != null && bounds.contains(mousePos);
        if (mousePos != null && bounds.contains(mousePos)) {
            Util.setCursor(QBubbles.getInstance().getPointerCursor());
            hovered = true;
        }
    }

    @Override
    public void unbindEvents() {
        QBubbles.getEventBus().unregister(this);
        eventsActive = false;

        if (hovered) {
            Util.setCursor(QBubbles.getInstance().getDefaultCursor());
            hovered = false;
        }
    }

    @Override
    public boolean eventsAreActive() {
        return eventsActive;
    }

    public void setBounds(Rectangle bounds) {
        this.bounds = bounds;
    }

    public void setHeight(int height) {
        this.bounds.height = height;
    }

    public void setWidth(int width) {
        this.bounds.width = width;
    }

    public void setX(int x) {
        this.bounds.x = x;
    }

    public void setY(int y) {
        this.bounds.y = y;
    }

    public boolean isHovered() {
        return hovered;
    }

    public boolean isPressed() {
        return pressed;
    }

    public Rectangle getBounds() {
        return bounds;
    }

    public RenderEventPriority getRenderEventPriority() {
        return renderEventPriority;
    }

    public void setRenderEventPriority(RenderEventPriority renderEventPriority) {
        this.renderEventPriority = renderEventPriority;
    }

    public Runnable getCommand() {
        return command;
    }

    public void setCommand(Runnable command) {
        this.command = command;
    }
}

File: src/main/kotlin/qtech/bubbles/gui/PauseButton.kt
@file:Suppress("unused")

package qtech.bubbles.gui

import qtech.bubbles.BubbleBlaster
import qtech.bubbles.common.interfaces.Listener
import qtech.bubbles.core.controllers.MouseController.Companion.instance
import qtech.bubbles.core.utils.categories.GraphicsUtils.drawCenteredString
import qtech.bubbles.event.MouseEvent
import qtech.bubbles.event.RenderEventPriority
import qtech.bubbles.event.SubscribeEvent
import qtech.bubbles.event.type.MouseEventType
import qtech.hydro.Border
import qtech.hydro.GraphicsProcessor
import qtech.bubbles.util.Util
import java.awt.*
import java.awt.geom.Rectangle2D
import java.util.*

class PauseButton protected constructor(x: Int, y: Int, width: Int, height: Int, var renderEventPriority: RenderEventPriority = RenderEventPriority.AFTER_FILTER) : Widget(), Listener {
    private val hash: Long = System.nanoTime()
    var bounds: Rectangle = Rectangle(x, y, width, height)
    var isHovered = false
        private set
    var isPressed = false
        private set
    var x: Int
        get() = bounds.x
        set(value) {
            bounds.x = value
        }
    var y: Int
        get() = bounds.y
        set(value) {
            bounds.y = value
        }
    var width: Int
        get() = bounds.width
        set(value) {
            bounds.width = value
        }
    var height: Int
        get() = bounds.height
        set(value){
            bounds.height = value
        }
    var location: Point
        get() = bounds.location
        set(value){
            bounds.location = value
        }
    var size: Dimension
        get() = bounds.size
        set(value){
            bounds.size = value
        }
    private var eventsActive = false
    var command: Runnable? = null
    var text: String? = null

    class Builder {
        private var _bounds = Rectangle(10, 10, 96, 48)
        private var _text = ""
        private var _renderPriority = RenderEventPriority.AFTER_FILTER
        private var _command = Runnable {}
        fun build(): PauseButton {
            val button = PauseButton(_bounds.x, _bounds.y, _bounds.width, _bounds.height, _renderPriority)
            button.text = _text
            button.command = _command
            return button
        }

        fun bounds(bounds: Rectangle): Builder {
            _bounds = bounds
            return this
        }

        fun bounds(x: Int, y: Int, width: Int, height: Int): Builder {
            _bounds = Rectangle(x, y, width, height)
            return this
        }

        fun text(text: String): Builder {
            _text = text
            return this
        }

        fun renderPriority(renderPriority: RenderEventPriority): Builder {
            _renderPriority = renderPriority
            return this
        }

        fun command(command: Runnable): Builder {
            _command = command
            return this
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as PauseButton
        return hash == that.hash
    }

    override fun hashCode(): Int {
        return Objects.hash(hash)
    }

    @SubscribeEvent
    fun onMouse(evt: MouseEvent) {
        if (evt.type === MouseEventType.PRESS) {
            if (bounds.contains(evt.point)) {
                isPressed = true
            }
        } else if (evt.type === MouseEventType.RELEASE) {
            if (bounds.contains(evt.point)) {
                command!!.run()
            }
            isPressed = false
            //        } else if (evt.getType() == MouseEventType.CLICK) {
        }
    }

    //    @SubscribeEvent
    //    public void onMouseMotion(MouseMotionEvent evt) {
    //        if (bounds.contains(evt.getParentEvent().getPoint())) {
    //            QBubbles.setCursor(Game.instance().getPointerCursor());
    //            hovered = true;
    //        } else {
    //            if (hovered) {
    //                QBubbles.setCursor(Game.instance().getDefaultCursor());
    //                hovered = false;
    //            }
    //        }
    //    }
    override fun paint(gp: GraphicsProcessor) {
//        hovered = MouseController.instance().getCurrentPoint() != null && bounds.contains(MouseController.instance().getCurrentPoint());
        val mousePos = instance().currentPoint
        if (mousePos != null) {
            if (bounds.contains(mousePos)) {
                Util.setCursor(BubbleBlaster.instance.pointerCursor)
                isHovered = true
            } else {
                if (isHovered) {
                    Util.setCursor(BubbleBlaster.instance.defaultCursor)
                    isHovered = false
                }
            }
        }
        val textColor: Color
        val oldStroke = gp.stroke()
        if (isPressed && instance().currentPoint != null && bounds.contains(instance().currentPoint!!)) {
            val old = gp.paint()
            val p = GradientPaint(0F, bounds.y.toFloat(), Color(0, 192, 255), 0f, (bounds.y + bounds.height).toFloat(), Color(0, 255, 192))
            gp.paint(p)
            val border = Border(1, 1, 1, 1)
            border.paint = Color(255, 255, 255, 128)
            border.paintBorder(gp, bounds.x, bounds.y, bounds.width, bounds.height)
            gp.paint(old)
            val border1 = Border(0, 0, 2, 0)
            border1.paint = p
            border1.paintBorder(gp, bounds.x, bounds.y, bounds.width, bounds.height)
            textColor = Color.white
        } else if (isHovered) {
            gp.stroke(BasicStroke(4.0f))
            val old = gp.paint()
            val shiftX = bounds.width.toDouble() * 2 * BubbleBlaster.ticks / (BubbleBlaster.TPS * 10)
            val p = GradientPaint(bounds.x + (shiftX.toFloat() - bounds.width), 0F, Color(0, 192, 255), bounds.x + shiftX.toFloat(), 0f, Color(0, 255, 192), true)
            gp.paint(p)
            gp.color(Color(255, 255, 255, 128))
            gp.shapeF(bounds)
            val border1 = Border(0, 0, 2, 0)
            border1.paint = p
            border1.paintBorder(gp, bounds.x, bounds.y, bounds.width, bounds.height)
            //            gg.draw(new Rectangle(bounds.x - 2, bounds.y - 2, bounds.width + 4, bounds.height + 4));
            gp.paint(old)
            textColor = Color(255, 255, 255)
        } else {
            gp.stroke(BasicStroke(1.0f))
            gp.color(Color(255, 255, 255, 128))
            gp.shapeF(bounds)
            //            Border border = new Border(1, 1, 1, 1);
//            border.setPaint(new Color(255, 255, 255, 128));
//            border.paintBorder(QBubbles.getInstance(), gg, bounds.x, bounds.y, bounds.width, bounds.height);
//            gg.draw(new Rectangle(bounds.x + 1, bounds.y + 1, bounds.width - 2, bounds.height - 2));
            textColor = Color(255, 255, 255, 128)
        }
        gp.stroke(oldStroke)
        val gg1 = gp.create(bounds.x + 1, bounds.y + 1, bounds.width - 2, bounds.height - 2)
        gg1.color(textColor)
        drawCenteredString(gg1, text!!, Rectangle2D.Double(0.0, 0.0, (bounds.width - 2).toDouble(), (bounds.height - 2).toDouble()), Font(BubbleBlaster.instance.font.name, Font.BOLD, 16))
        gg1.dispose()
    }

    override fun destroy() {}
    fun tick() {}
    override fun bindEvents() {
        BubbleBlaster.eventBus.register(this)
        eventsActive = true
        val mousePos = instance().currentPoint
        if (mousePos != null && bounds.contains(mousePos)) {
            Util.setCursor(BubbleBlaster.instance.pointerCursor)
            isHovered = true
        }
    }

    override fun unbindEvents() {
        BubbleBlaster.eventBus.unregister(this)
        eventsActive = false
        if (isHovered) {
            Util.setCursor(BubbleBlaster.instance.defaultCursor)
            isHovered = false
        }
    }

    override fun eventsAreBound(): Boolean {
        return eventsActive
    }

}